src/wifipaf/WiFiPAFEndPoint.cpp

```diff
@@ -797,78 +787,88 @@ SequenceNumber_t WiFiPAFEndPoint::AdjustRemoteReceiveWindow(SequenceNumber_t las
     }
     // Always has header flag
     SnOffset += kTransferProtocolHeaderFlagsSize;
-    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kManagementOpcode)) // Has Mgmt_Op
+    if (rxFlags.Has(WiFiPAFTP::HeaderFlags::kManagementOpcode)) // Has Mgmt_Op
     {
         SnOffset += kTransferProtocolMgmtOpSize;
     }
-    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kFragmentAck)) // Has ack
+    if (rxFlags.Has(WiFiPAFTP::HeaderFlags::kFragmentAck)) // Has ack
     {
         SnOffset += kTransferProtocolAckSize;
     }
-    pSn    = pHead + SnOffset;
-    seqNum = *pSn;
-
-    return CHIP_NO_ERROR;
+    static_assert(std::is_same<SequenceNumber_t, uint8_t>::value, "Type of SequenceNumber_t is unexpected.");
+    err = readerSn.Skip(SnOffset).ReadBytes(&seqNum, 1).StatusCode();
+    SuccessOrExit(err);
+exit:
+    return err;
 }
 
-CHIP_ERROR WiFiPAFEndPoint::DebugPktAckSn(const PktDirect_t PktDirect, Encoding::LittleEndian::Reader & reader, uint8_t * pHead)
+CHIP_ERROR WiFiPAFEndPoint::DebugPktAckSn(const PktDirect_t pktDirect, Encoding::LittleEndian::Reader & reader)
 {
 #ifdef CHIP_WIFIPAF_END_POINT_DEBUG_LOGGING_ENABLED
-    BitFlags<WiFiPAFTP::HeaderFlags> rx_flags;
+    auto readerHdr = reader;
+    auto readerSn  = reader;
+    BitFlags<WiFiPAFTP::HeaderFlags> rxFlags;
     CHIP_ERROR err;
-    uint8_t * pAct = nullptr;
+    bool hasAck = false;
+    SequenceNumber_t ackNum, snNum;
     char AckBuff[4];
-    uint8_t * pSn;
-    uint8_t SnOffset = 0;
+    size_t SnOffset = 0;
 
-    err = reader.Read8(rx_flags.RawStorage()).StatusCode();
+    err = readerHdr.Read8(rxFlags.RawStorage()).StatusCode();
     SuccessOrExit(err);
-    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kHankshake))
+    if (rxFlags.Has(WiFiPAFTP::HeaderFlags::kHandshake))
     {
         // Handkshake message => No ack/sn
         return CHIP_NO_ERROR;
     }
     // Always has header flag
     SnOffset += kTransferProtocolHeaderFlagsSize;
-    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kManagementOpcode)) // Has Mgmt_Op
+    if (rxFlags.Has(WiFiPAFTP::HeaderFlags::kManagementOpcode)) // Has Mgmt_Op
     {
         SnOffset += kTransferProtocolMgmtOpSize;
     }
-    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kFragmentAck)) // Has ack
+    if (rxFlags.Has(WiFiPAFTP::HeaderFlags::kFragmentAck)) // Has ack
     {
-        pAct = pHead + kTransferProtocolHeaderFlagsSize;
+        auto readerAck = reader;
+        err            = readerAck.Skip(kTransferProtocolHeaderFlagsSize).ReadBytes(&ackNum, 1).StatusCode();
+        SuccessOrExit(err);
+        hasAck = true;
         SnOffset += kTransferProtocolAckSize;
     }
-    pSn = pHead + SnOffset;
-    if (pAct == nullptr)
+    err = readerSn.Skip(SnOffset).ReadBytes(&snNum, 1).StatusCode();
+    SuccessOrExit(err);
+    if (hasAck == false)
     {
         strcpy(AckBuff, "  ");
     }
     else
     {
-        snprintf(AckBuff, sizeof(AckBuff), "%02hhu", *pAct);
+        snprintf(AckBuff, sizeof(AckBuff), "%02hhu", ackNum);
```

Comment:
Consider using a StringBuilder to manage storage automatically without sizeof and similar (it also truncates nicely)

Reply:
I will change it